Save and parse Arkane species with all statmech properties using YAML format

alongd opened the pull request:
We'd like to be able to conveniently use previously calculated statmech data in Arkane (prev. Cantherm). The purpose is to have an online repository of such files, from which users can pick the desired species at the appropriate level of theory (if available). This PR sets the ground by allowing Arkane to save and parse such files.

mliu49 commented:
If this is going to be a brand new file type, I strongly suggest using an established format such as yaml, which you can parse using pyyaml to give a python dict. Using exec on user provided input files is a security risk since we could be executing arbitrary code. It would be best to convert all of our input files and database files to not require exec, but it'll take time.

Regarding the idea of having a database of statmech info, it would again be best if it's not a plain-text database like RMG-database is currently. Kehang's thermo central database uses mongoDB, which is an option, especially since it seems like it would overlap or be a part of this future statmech database. There are also tons of other options, but we should definitely put some thought into the database design.

alongd commented:
I changed the format into `.yml`.
Still, we need to use `eval()` (instead of `execute()`), because loading the YML file only gives dictionaries of strings, and the objects must be reconstructed.
I suggest we add more thorough checks to the constructor functions I added to common.py that the string we're evaluating is indeed what we think it is (contains several key words, under a certain maximum length, etc.). Other suggestions are welcomed!

mliu49 commented on rmgpy/cantherm/data/HSOO.yml:
```diff
@@ -0,0 +1,43 @@
+frequency_scale_factor: 0.975
+label: HSOO
+level_of_theory: CCSD(T)-F12a/cc-pVTZ-F12//B3LYP/6-311G(2d,d,p), rotors at B3LYP/6-311G(2d,pd)
+model_chemistry: CCSD(T)-F12/cc-pVTZ-F12
```
I think the `level_of_theory` and `model_chemistry` may be a bit confusing/repetitive. This would require some refactoring of existing code, but you could consider renaming `model_chemistry` to `level_of_theory_energy`, then create two new attributes called `level_of_theory_geometry` and `level_of_theory_rotors`.

alongd replied:
I agree that it's a bit confusing this way, but I'm not sure that `model_chemistry` and `level_of_theory_energy` are interchangeable. Since `model_chemistry` is being used by Cantherm to apply the respective bond and atom corrections, users could specify a "close enough" `model_chemistry` to the actual `level_of_theory_energy` they used. I'm not saying it's a good thing to do, but might happen. I want to report here both, so first we know what real `level_of_theory_energy` was used, but also which type of corrections were implemented.

Perhaps a good approach to solve this confusion would be to only dump `model_chemistry` into the YAML file if it's different than the energy level of theory.

mliu49 commented on rmgpy/cantherm/input.py:
```diff
@@ -124,8 +139,16 @@ def species(label, *args, **kwargs):
+            spec.molecule = [structure]
         spec.conformer = Conformer(E0=E0, modes=modes, spinMultiplicity=spinMultiplicity, opticalIsomers=opticalIsomers)
+        if molecularWeight is None:
+            if is_pdep():
```
Does this work if the species block comes before the pdep block?

Also, if you refactor the molecular weight implementation based on my other comment, would that eliminate the need for this if block?

alongd replied:
modified

mliu49 commented on rmgpy/cantherm/input.py:
```diff
@@ -388,27 +411,111 @@ def loadInputFile(path):
+    """
+    Load the species with all statMech data from the .yml file
+    """
+    global jobList
```
It doesn't seem like you use jobList in this function.

alongd replied:
removed

mliu49 commented on rmgpy/cantherm/input.py:
```diff
@@ -388,27 +411,111 @@ def loadInputFile(path):
+                    coordinates = np.ndarray(shape=(len(number), 3))
+                    for i, coorlist in enumerate(job.species.conformer.coordinates.value_si):
+                        coordinates[i] = [coor for coor in coorlist]
+                    mol = Molecule().fromXYZ(atomicNums=number, coordinates=coordinates)
```
This section seems oddly coded. Was it done intentionally? It seems like the following would be sufficient:
```
number = job.species.conformer.number.value_si.tolist()
coordinates = job.species.conformer.coordinates.value_si.tolist()
```

mliu49 replied:
Oh, I misinterpreted it, I thought initially that the conformer attributes were `np.ndarray` objects, and you were converting them into lists.

If the conformer attributes are `np.ndarray`, it seems you could use them directly. If they're lists, then couldn't you use `np.array(list_input)` to convert it to an `np.ndarray`?

mliu49 commented on rmgpy/cantherm/statmech.py:
```diff
@@ -200,6 +206,12 @@ def load(self):
+        filename, file_extension = os.path.splitext(self.path)
+        if 'yml' in file_extension or 'yaml' in file_extension:
+                self.load_species_from_database = True
+                return
```
It seems like you could just call `load_species_from_database_file` here instead of `main.py`, and it would also eliminate the need for the `load_species_from_database` attribute.

alongd replied:
I agree, will change accordingly

mliu49 commented on rmgpy/cantherm/thermo.py:
```diff
@@ -188,6 +188,30 @@ def save(self, outputFile):
 
+        return chem_string
+
+    def save_species_to_database_file(self, path):
```
I think this may be more appropriate as a method of `CanthermSpecies`. Let me know what you think.

alongd replied:
I agree

alongd commented on rmgpy/cantherm/common.py:
```diff
@@ -43,6 +63,296 @@ def checkConformerEnergy(Vlist,path):
+                        ' kJ/mol. This can cause significant errors in your computed rate constants. ')
+
+
+def determine_molecular_weight(species):
```
Thanks for the comment!
I'll implement it in the Species `__init__()`:
```python
def __init__(self, ..., molecularWeight=None, ...):
        if self.molecularWeight is None:
            self._molecularWeight = quantity.Mass(self.molecule[0].getMolecularWeight(), 'kg/mol')
        else:
            self.molecularWeight = molecularWeight
```
Shall I still change the syntax of `molecularWeight`? I can leave it untouched so #1329 will have less conflicts

alongd replied:
On second though, your recommendation seems better if for some reason an empty Species object is first created, then a structure is assigned.

alongd commented on rmgpy/cantherm/common.py:
```diff
@@ -43,6 +63,296 @@ def checkConformerEnergy(Vlist,path):
+            d['thermo'] = species.thermo
+        d['chemkin_thermo_string'] = chemkin_thermo_string
+
+        self.species_dictionary = d
```
I agree.
I think I still need something like the update() method, though, to only append the species attributes to this object (the object is initiated with non-species-specific attributes such as `author` and `level_of_theory` which are kept)

alongd commented on rmgpy/cantherm/data/HSOO.yml:
```diff
@@ -0,0 +1,43 @@
+transport_data: TransportData(epsilon=(3625.11,'J/mol'), sigma=(5.7,'angstrom'))
+use_atom_corrections: true
+use_bond_corrections: false
+use_hindered_rotors: true
```
At least `use_bond_corrections` is crucial, the rest are just "nice to have".

`use_bond_corrections`: This might impact the decision of the user of whether or not to use this file. If the user wants to use this species in a PES with other species calculated at different levels of theory, it is important/recommended to use bond corrections. On the other hand, if all species were calculated at the same level, it is better practice not to use bond corrections (which don't capture all bond types, and are often ambiguous for TSs).

`use_hindered_rotors`: This could also be directly inferred from the Conformer object. However, for a quick assessment, it's nice to have it explicitly written.

`use_atom_corrections`: Can be inferred from the `atom_energies` attribute, I'm OK with removing it since using this feature is less common.

alongd commented:
@mliu49, Thanks for all the valuable comments!
I addressed them in fixup commits. At some point I didn't bother correctly naming the fixup commits due to the relocation of methods between files, I think I'll re-commit it all instead of squashing.
Let me know if you have any additional comments.

alongd commented:
@mliu49, can I squash this branch?

mliu49 commented:
Yeah, you can go ahead and squash. I guess this might get replaced by the new method though, it that turns out ok.

mliu49 commented on rmgpy/rmg_obj.py:
```diff
@@ -0,0 +1,52 @@
+###############################################################################
+import numpy as np
+
+class RMG_Obj(object):
```
A couple comments on naming:
- Class names should be CamelCase, though I think we can make an exception for RMG. I would probably call this `RMGObject`.
- Module names are ideally one word. Perhaps something like `object`, `base`, or `common`?
- Variable names should be lowercase. I think `D` should be renamed to something else, like `output` or even `output_dict`.

Also, could you end the file with a newline?

mjohnson541 commented:
They're our own objects right?  So can't we just make them inherit from RMGObject?

alongd commented on rmgpy/rmgobject.py:
```diff
@@ -0,0 +1,51 @@
+    def __init__(self):
+        pass
+    def as_dict(self):
+        D = self.__dict__.copy()
```
Thanks for this addition, @mjohnson541 !
I think that `self.__dict__` [contains methods as well](https://stackoverflow.com/questions/61517/python-dictionary-from-an-objects-fields), and we should first check `if not callable(value)` before adding entries to the output dictionary.

alongd commented:
I rewrote this branch using the `as_dict()` and `make_object()` methods we discussed for YAML loading and parsing.
Thanks @mjohnson541 and @mliu49 for the discussion and help with writing the methods.
The branch is now ready for another review cycle.

I'll note that since cytonized classes cannot inherit from non-cytonized classes, and vice versa, I made two parent classes: `RMGObject` which is cytonized, and `RMGPYObject` which isn't.
Quantity (Scalar & Array) and the thermo classes (NASA & Wilhoit) were not set to inherit from the parent objects. I think it's OK for quantity, but we could theoretically modify the way attributes are saved in the thermo classes so that they too can inherit from `RMGObject`. But I didn't think it was that important when I implemented it.

mliu49 commented:
Are you sure that Python classes can't subclass Cython classes? It should be possible according to the [Cython documentation](https://media.readthedocs.org/pdf/cython/stable/cython.pdf) (section 3.5).

alongd commented:
Thanks for the guidance, @mliu49 .
Now only one RMGObject class (cythonized) is implemented.

alongd commented:
@mliu49, the MCH test has timed out in this PR. I restarted the `push` tests, and got the same behaviour. Could you take a look?

mliu49 commented:
I wouldn't expect this branch to affect RMG jobs, so I'm not sure what would cause it to take so long. My only suggestion would be to run it locally to see what it does.

alongd commented:
It ran fine locally within a reasonable time (we had some rates violating the collision limit, but that should be meaningless for the test)
```Saving current model core to Chemkin file...
Chemkin file contains 323 reactions.
Saving current model to verbose Chemkin file...
Chemkin file contains 323 reactions.
Saving current model core and edge to Chemkin file...
Chemkin file contains 2197 reactions.
Saving current core and edge to verbose Chemkin file...
Chemkin file contains 2197 reactions.
Updating RMG execution statistics...
    Execution time (DD:HH:MM:SS): 00:00:06:11
    Memory used: 989.93 MB

Making seed mechanism...
Performing final model checks...


Warning: 4 CORE reactions violate the collision rate limit!
See the 'collision_rate_violators.log' for details.



MODEL GENERATION COMPLETED

The final model core has 81 species and 322 reactions
The final model edge has 654 species and 1872 reactions

RMG execution terminated at Thu Jul 12 22:30:49 2018
```

alongd commented:
@mliu49 and @mjohnson541, I think this branch is ready for another review cycle, getting us closer to a YAML RMG future

alongd commented:
@mliu49, I rebased the Arkane YAML file PR.
All comments from above were implemented. I think we should be good to go, unless you have additional comments. Please take a look whenever you got time. Thanks!

mliu49 commented on arkane/statmech.py:
```diff
@@ -175,29 +176,41 @@ def __init__(self, species, path):
+        _, file_extension = os.path.splitext(path)
+        if file_extension in ['.yml', '.yaml']:
+            if TS:
+                raise ValueError('Loading transition states from a YAML file is still unsupported.')
```
This could be a `NotImplementedError`.

alongd commented on rmgpy/rmgobject.pyx:
```diff
@@ -0,0 +1,113 @@
+# cython: embedsignature=True, cdivision=True
```
Thanks for alerting me, it's a result of copy-pasting another pyx file as template. Removed

alongd commented on rmgpy/rmgobject.pyx:
```diff
@@ -0,0 +1,113 @@
+            elif not isinstance(val, (int, float, str, dict)):
+                if isinstance(val, np.ndarray):
+                    output_dict[key] = val.tolist()
+                elif val is not None:
```
You're right, it cannot be `None`. But it doesn't have to be an RMGObject either: some of our classes have tailored `as_dict` and `make_object` methods and do not inherit from RMGObject (e.g., scalar and array quantities). I'll try removing the condition.

alongd commented on rmgpy/quantity.py:
```diff
@@ -395,8 +462,12 @@ def getValue(self):
     def setValue(self, v):
-        self.value_si = numpy.array(v) * self.getConversionFactorToSI()
+        if isinstance(v, float):
+            self.value_si = numpy.array([v]) * self.getConversionFactorToSI()
```
I'm not sure (it was written a while ago), but I think I had an instance like that when playing around with this branch. Maybe an ArrayQuantity with only one entry saved to YAML (such as a biatomic molecule's frequency) is read as a float? I'm also OK with removing this addition, and adding it back only if we encounter a related bug.

alongd commented on arkane/common.py:
```diff
@@ -31,12 +31,233 @@
+        full_path = os.path.join(path, filename)
+        with open(full_path, 'w') as f:
+            yaml.dump(data=self.as_dict(), stream=f, canonical=False)
+        # remove empty lines from the file (multi-line strings have excess new line brakes for some reason):
```
Thanks for the link above to the solution on stackoverflow re extra line breaks. The solution there uses a string representer, but we no longer use those since we don't `evaluate` any more. I'll implement your suggestion here.

alongd commented:
@mliu49, thanks for the helpful comments!
I addressed them all either in the comments above or as fixup commits.
Let me know if there's anything else you think we should change

alongd commented:
rebased. @mliu49, let me know when I can squash.

alongd commented:
@mliu49, let me know when to squash the fixups to finalize the PR

mliu49 commented on arkane/input.py:
```diff
@@ -214,12 +214,11 @@ def species(label, *args, **kwargs):
-                                 " cannot be reconstructed.".format(spec.label))
+                if spec.molecularWeight is None and is_pdep(jobList):
+                    # If one of the jobs is pdep and no molecular weight is given or calculated, raise an error
+                    raise ValueError("No molecularWeight was entered for species {0}. Since a structure wasn't given"
```
Hmm, my suggestion might not have been correct, since this now misses the case where a structure is not provided. One possibility:
```
if molecularWeight is not None:
    spec.molecularWeight = molecularWeight
elif spec.molecularWeight is None and is_pdep(jobList):
    raise ValueError()
```
There shouldn't be any need to check whether structure exists, because the the Species.molecularWeight property checks if it has a molecule attribute, which will be empty if structure was empty.

mliu49 commented on rmgpy/rmgobject.pyx:
```diff
@@ -1,10 +1,11 @@
-# cython: embedsignature=True, cdivision=True
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
```
This actually isn't necessary either, since this is technically not a python file.

mliu49 commented on arkane/qchem.py:
```diff
@@ -273,12 +273,13 @@ def loadEnergy(self, frequencyScaleFactor=1.):
         with open(self.path, 'r') as f:
-            for line in f:
+            lines = f.readlines()
+            for line in lines:
```
The original implementation is preferable because it treats the file as an iterator, so only one line is loaded at a time. If you use readlines(), then the entire file has to be loaded into memory.

alongd replied:
OK, I think this isn't related to this PR (sorry) so I removed it, we could discuss it at another time (I think it might be necessary for running Arkane through ARC). Thanks for bringing this up, I removed the commit.
